File: Data Structure/6. Queue/Priority Queue/priorityQueue.cpp
// array implementation of priority queue
#include <iostream>
#include <vector>
using namespace std;

class PriorityQueue {
  // vector in sorted order, from max at 0 to min at size-1
  int maxSize;
  vector<int> queue;
  int nItems;
  
  public:
    // constructor
    PriorityQueue(int siz) {
      maxSize = siz;
      nItems = 0; 
      queue.resize(maxSize);
    }

    // insert item to the priority queue
    void insert(int item) {
      // if empty, insert at index 0
      if (nItems == 0)
        queue[nItems++] = item;
      else {
        // start at the end of the queue
        int i = nItems-1;
        for (i; i >= 0; i--) {
          // if new item larger than current element
          if (item > queue[i]) 
            queue[i+1] = queue[i]; // shift upward
          else
            break; // finish shifting
        }
        queue[i+1] = item; // insert new item
        nItems++; // increament size counter
      }
    }

    // remove minimum item
    int remove() {
      return queue[--nItems];
    }

    // get minimum value
    int getMin() {
      return queue[nItems-1];
    }

    // check if it is empty
    bool isEmpty() {
      return nItems == 0;
    }

    // get number of item in the queue
    int getSize() {
      return nItems;
    }

    // check if the queue is full
    bool isFull() {
      return nItems == maxSize;
    }
};

int main() {
  cout << "Testing priority queue class\n\n";

  PriorityQueue pq(5);
  cout << "- Created a priority queue object with size 5\n";
  cout << "Queue Size is: " << pq.getSize() << endl;

  pq.insert(30);
  pq.insert(29);
  pq.insert(28);
  pq.insert(31);
  pq.insert(32);
  pq.insert(33);
  cout << "- Inserted values into priority queue\n";
  cout << "Queue Size is: " << pq.getSize() << endl;

  while (!pq.isEmpty()) {
    int top = pq.remove();
    cout << top << " ";
  }
  cout << "\n- Cleared values into priority queue\n";
  cout << "Queue Size is: " << pq.getSize() << endl;
}

File: Data Structure/6. Queue/Priority Queue/priorityQueue.py
# list implementation of priority queue
class PriorityQueue:
  # constructor
  def __init__(self, siz):
    self.nItems = 0
    self.queue = [0 for i in range(siz+1)]

  # insert operation
  def insert(self, item):
    # if empty, insert at index 0
    if self.nItems == 0:
      self.queue[self.nItems] = item
      self.nItems += 1
    else:
      # start the end of the queue
      i = self.nItems-1
      while i >= 0:
        # if new item larger than current element
        # move the smaller current item to the back
        if item > self.queue[i]:
          self.queue[i+1] = self.queue[i] # shift upward
        else:
          break
        i -= 1
      self.queue[i+1] = item # insert new item
      self.nItems += 1 # increment size counter
    
  # remove minimum item
  def remove(self):
    self.nItems -= 1
    return self.queue[self.nItems]
  
  # get minimum value
  def getMin(self):
    return self.queue[self.nItems-1]

  # check if it is empty
  def isEmpty(self):
    return self.nItems == 0
  
  # get number of item in the queue
  def getSize(self):
    return self.nItems

# ================================
# testing the PriorityQqueue class
# ================================

print("Testing priority queue class\n")

# create priority queue with size 5
pq = PriorityQueue(5)
print("- Created a priority queue object with size 5\n")
print("Queue size is: " + str(pq.getSize()) + "\n\n")

# insert values into the queue
pq.insert(30)
pq.insert(29)
pq.insert(28)
pq.insert(31)
pq.insert(32)
pq.insert(33)
print("- Inserted values into priority queue\n")
print("Queue size is: " + str(pq.getSize()) + "\n\n")

# clear the queue
while not pq.isEmpty():
  top = pq.remove()
  print(str(top) + " "),
print
print("- Cleared values for priority queue\n")
print("Queue size is: " + str(pq.getSize()) + "\n\n")